### PR TITLE
feat: Faster UDP/IO on Apple platforms

### PR DESCRIPTION
**Pull request by larseggert**
This uses Apple's private [`sendmsg_x`](https://github.com/apple-oss-distributions/xnu/blob/94d3b452840153a99b38a3a9659680b2a006908e/bsd/sys/socket.h#L1457-L1487) and [`recvmsg_x`](https://github.com/apple-oss-distributions/xnu/blob/94d3b452840153a99b38a3a9659680b2a006908e/bsd/sys/socket.h#L1425-L1455) system calls for multi-packet UDP I/O.

CC @mxinden

### CODE REVIEW

**Comment by Ralith** [comment]
> Is there interest in seeing TX support via sendmsg_x?

We found there wasn't much performance benefit, and was considerable difficulty taking advantage of, `sendmmsg`-style batching. IIRC the `_x` functions on macOS have more to offer than that, though. Will this unblock segmentation offload or other incidental optimizations?

File: quinn-udp/build.rs
Comment (Ralith) [inline quinn-udp/build.rs:8]: Bindgen is a pretty big and unstable dependency. Would it be simpler to commit the generated bindings?

**Comment by larseggert** [comment]
Bench on `main`:
```
test large_data_10_streams  ... bench:  27,558,791 ns/iter (+/- 13,459,810) = 380 MB/s
test large_data_1_stream    ... bench:  24,324,266 ns/iter (+/- 19,219,937) = 43 MB/s
test small_data_100_streams ... bench:  19,437,900 ns/iter (+/- 20,065,941)
test small_data_1_stream    ... bench:  11,465,128 ns/iter (+/- 8,699,934)
```

Bench with this PR:
```
test large_data_10_streams  ... bench:  28,829,216 ns/iter (+/- 15,924,956) = 363 MB/s
test large_data_1_stream    ... bench:  14,354,999 ns/iter (+/- 20,039,122) = 73 MB/s
test small_data_100_streams ... bench:  14,061,741 ns/iter (+/- 17,311,517)
test small_data_1_stream    ... bench:  19,194,441 ns/iter (+/- 5,012,070)
```

Surprised that `large_data_10_streams` and `small_data_1_stream` are slower...

**Comment by larseggert** [comment]
We've also found on neqo that multi-packet RX without multi-packet TX has limited benefits, since the RX batch size will be very small.

**Comment by larseggert** [comment]
I added `sendmsg_x` support, mostly to see what the performance difference would be. But it seems that none of the benches or tests call `send` with a `Transmit` struct where `segment_size` is not `None`?

**Comment by mxinden** [comment]
> A targeted quinn-udp benchmark might be more useful.

How about using the `throughput.rs` benchmark @larseggert?

https://github.com/quinn-rs/quinn/blob/main/quinn-udp/benches/throughput.rs

**Comment by larseggert** [comment]
With @mxinden's benchmark. Baseline:
```
gso_true/throughput     time:   [58.076 ms 58.230 ms 58.387 ms]
                        thrpt:  [171.27 MiB/s 171.73 MiB/s 172.19 MiB/s]
```

Only `sendmsg_x`:
```
gso_true/throughput     time:   [15.143 ms 15.189 ms 15.236 ms]
                        thrpt:  [656.35 MiB/s 658.36 MiB/s 660.37 MiB/s]
                 change:
                        time:   [-74.028% -73.915% -73.808%] (p = 0.00 < 0.05)
                        thrpt:  [+281.80% +283.36% +285.04%]
                        Performance has improved.
```

Both `sendmsg_x` and `recvmsg_x`:
```
gso_true/throughput     time:   [12.632 ms 12.682 ms 12.731 ms]
                        thrpt:  [785.46 MiB/s 788.53 MiB/s 791.61 MiB/s]
                 change:
                        time:   [-78.321% -78.221% -78.112%] (p = 0.00 < 0.05)
                        thrpt:  [+356.88% +359.16% +361.27%]
                        Performance has improved.
```

File: quinn-udp/benches/throughput.rs
Comment (djc) [inline quinn-udp/benches/throughput.rs]: `slives` is probably not what you were going for?

File: quinn-udp/src/unix.rs
Comment (djc) [inline quinn-udp/src/unix.rs:393]: Nit: please flip the condition and make this an early return.

File: quinn-udp/src/unix.rs
Comment (djc) [inline quinn-udp/src/unix.rs:392]: I think we can get rid of the commented out debug cruft now?

File: quinn-udp/src/unix.rs
Comment (djc) [inline quinn-udp/src/unix.rs:521]: Nit: suggest spelling it like this:

```rust
match n {
    -1 => {
         ..
    }
    n => break n,
}
```

File: quinn-udp/src/unix.rs
Comment (larseggert) [reply quinn-udp/src/unix.rs:392]: This always returns `Ok(())`, just as in the BSD version below. (Should both of these actually return an Error in this case?)

File: quinn-udp/src/unix.rs
Comment (mxinden) [inline quinn-udp/src/unix.rs:386]: Nit: It is not idiomatic to execute a side-effect in an `Iterator::map`.

> If you’re doing some sort of looping for a side effect, it’s considered more idiomatic to use [for](https://doc.rust-lang.org/stable/book/ch03-05-control-flow.html#looping-through-a-collection-with-for) than map().

https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#method.map

While this might seem dogmatic, I do find reasoning about an Iterator chain much easier, when being able to rely on the convention of no-side-effects-in-map.

Instead I suggest either using a standard `for` loop here, or `Iterator::for_each`, updating `cnt` on each iteration.

File: quinn-udp/src/unix.rs
Comment (mxinden) [reply quinn-udp/src/unix.rs:391]: > Nit: please flip the condition and make this an early return.

Agreed with @larseggert. I suggest keeping as is, as both paths (`== -1` and `>-1`) have the same return value `Ok(())` and given that it mirrors the BSD implementation below.

> Should both of these actually return an Error in this case?

In case we do want to return an error here, I suggest doing this as part of https://github.com/quinn-rs/quinn/issues/1971 across all platforms (e.g. as you mention BSD) and not in this pull request.

File: quinn-udp/src/unix.rs
Comment (mxinden) [inline quinn-udp/src/unix.rs:825]: I think it is worth documenting here, that UDP Datagram segmentation is not offloaded to the OS or even better the NIC, but instead done here in user space in `fn send()` and then passed to the OS as individual `iovec`s.

**Review by mxinden — COMMENTED** [review]
Impressive results. Great to see the MacOS `_x` syscalls work for QUIC UDP IO.

File: quinn-udp/benches/throughput.rs
Comment (Ralith) [inline quinn-udp/benches/throughput.rs:36]: This change seems to be confusing GRO with multi-message receive. GRO batches are passed to userspace as a single packet; that's why we used a single large buffer. By switching to many small buffers, you've actually disabled GRO.

**Review by Ralith — DISMISSED** [review]
Can we enable real GSO/GRO using these interfaces?

**Comment by larseggert** [comment]
No. They are the equivalent of the `mmsg` Linux calls. AFAIK Apple doesn't have GSO/GRO via the socket interface.

File: quinn-udp/benches/throughput.rs
Comment (larseggert) [reply quinn-udp/benches/throughput.rs:36]: Apple doesn't **do** GSO/GRO, at least not via the socket interface. The `_x` calls are the semantic equivalent of the Linux `mmsg` calls. I'm borrowing the GSO/GRO helpers set up multi-message TX/RX in the bench.

File: quinn-udp/src/unix.rs
Comment (larseggert) [reply quinn-udp/src/unix.rs:386]: Can do either. What do the maintainers prefer?

File: quinn-udp/src/cmsg/unix.rs
Comment (djc) [inline quinn-udp/src/cmsg/unix.rs:53]: I guess this comment isn't relevant for the macOS implementation?

File: quinn-udp/src/unix.rs
Comment (djc) [inline quinn-udp/src/unix.rs:32]: Looks like so far we've imported symbols from `std::ffi` directly, keeping `libc` symbols as prefixed. I'd be in favor of sticking with that style.

File: quinn-udp/src/unix.rs
Comment (djc) [inline quinn-udp/src/unix.rs:43]: Are you contributing this to upstream libc? (We can carry it until it gets released.)

File: quinn-udp/src/unix.rs
Comment (djc) [reply quinn-udp/src/unix.rs:386]: Yes, agreed that a `for` loop would be better here.

File: quinn-udp/src/unix.rs
Comment (djc) [reply quinn-udp/src/unix.rs:391]: > Agreed with @larseggert. I suggest keeping as is, as both paths (`== -1` and `>-1`) have the same return value `Ok(())` and given that it mirrors the BSD implementation below.

I don't find this a very strong argument. This saves some rightward drift for the error handling code and duplicating the `Ok(())` return value doesn't seem like a particularly big deal here. @Ralith opinions?

File: quinn-udp/src/unix.rs
Comment (mxinden) [inline quinn-udp/src/unix.rs:915]: `quinn-udp` **does not** expose `sendmmsg` style sends, i.e. only supports multiple sends via GSO from a single buffer with equal size segments. Thus the above hack of returning `BATCH_SIZE` on `max_gso_segments`, despite the fact that MacOS does not support GSO, while not ideal, seems reasonable. The large multi-segment datagram received from the user is segmented in user-space and then passed to MacOS as individual `iovec`s, `sendmmsg` style.

`quinn-udp` **does** expose `recvmmsg` style reads, i.e. supports receiving multiple datagrams both in separate `iovec`s and via GRO. Given that MacOS does not support GRO, why return `BATCH_SIZE` here and not `1`?

```suggestion
        1
```

File: quinn-udp/src/unix.rs
Comment (larseggert) [reply quinn-udp/src/unix.rs:43]: I wasn't planning on, since these are private APIs.

File: quinn-udp/src/unix.rs
Comment (larseggert) [reply quinn-udp/src/unix.rs:915]: So that `UdpSocketState::gro_segments` is initialized to something other than 1? What am I missing?

File: quinn-udp/src/unix.rs
Comment (djc) [reply quinn-udp/src/unix.rs:46]: Ohh, had missed that. I guess they might not take them. So given that these are documented as

> NOTE: This a private system call, the API is subject to change.

Should we make this an opt-in feature? Does private here only mean "subject to change" or also "will prevent your app from being allowed in the App Store"?

Comment (larseggert) [reply quinn-udp/src/unix.rs:46]: I don't know. I do know that these calls have been there for 10+ years. I can put them behind a feature if you prefer.

Comment (larseggert) [reply quinn-udp/src/unix.rs:46]: Maybe @ekinnear knows...

File: quinn-udp/src/unix.rs
Comment (mxinden) [reply quinn-udp/src/unix.rs:912]: > So that `UdpSocketState::gro_segments` is initialized to something other than 1?

Why does it need to be initialized to something other than `1`?

If `UdpSocketState::gro_segments` is `1`, a user can still provide multiple buffers to `UdpSocketState::recv` and, on the MacOS code-path, each of those buffers can be passed to the OS to be filled in a `recvmmsg` style.

Comment (mxinden) [reply quinn-udp/src/unix.rs:912]: Put differently, the value of `UdpSocketState::gro_segments` only determines, how many segments a user can receive in one `iovec`, not how many `iovec`s a user can pass to `UdpSocketState::recv()`.

In the case of quinn as a user, it simply determines how large each `iovec` should be, i.e. how many segments each `iovec` should be able to fit:

https://github.com/quinn-rs/quinn/blob/9a656d2d0b3031390e71a24258b5432115cfa381/quinn/src/endpoint.rs#L745-L750

Note that `recv_buf` here represents all `iovec`s in sequence passed to `UdpSocketState::recv()`.

(Quinn maintainers, please correct me if I am missing something.)

File: quinn-udp/benches/throughput.rs
Comment (Ralith) [inline quinn-udp/benches/throughput.rs:46]: Is this relevant to throughput testing? We already cover the ability to send/receive ECN in the unit tests.

File: quinn-udp/benches/throughput.rs
Comment (Ralith) [reply quinn-udp/benches/throughput.rs]: Sure, but this isn't an Apple-specific test; it runs on all platforms quinn-udp supports, some of which do have real GRO/GSO. This change will undermine performance testing on those platforms.

Comment (larseggert) [reply quinn-udp/benches/throughput.rs]: This changed to
```rust
    let mut meta = vec![RecvMeta::default(); BATCH_SIZE];
```
in the meantime; is this still a concern?

(Because it seems that for `recvmmsg` platforms, you already needed that change?)

File: quinn-udp/benches/throughput.rs
Comment (larseggert) [reply quinn-udp/benches/throughput.rs:46]: Sorry this snuck in from local testing.

File: quinn-udp/benches/throughput.rs
Comment (mxinden) [reply quinn-udp/benches/throughput.rs]: > This changed to
> 
> ```rust
>     let mut meta = vec![RecvMeta::default(); BATCH_SIZE];
> ```

:+1: Thanks.

> in the meantime; is this still a concern?
> 
> (Because it seems that for `recvmmsg` platforms, you already needed that change?)

I initially wrote this benchmark (https://github.com/quinn-rs/quinn/pull/1915) to test GSO/GRO only. I am in favor of your addition, namely to leverage `recvmmsg` as well, i.e. pass more than one `iovec` to the OS to read into. Thanks @larseggert. I assume that on MacOS this benchmark now returns the same result for both `gro_result` `true` and `false`, correct?

To avoid any confusion, other than MacOS and iOS, I don't think there are any supported platforms that do `recvmmsg` but not `GRO`.

Linux / Android does [recvmmsg](https://man7.org/linux/man-pages/man2/recvmmsg.2.html) with GRO,  and Windows does [WSARecvMsg](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/legacy/ms741687(v=vs.85)) (equivalent to `recvmmsg`) with GRO. The BSDs do neither `recvmmsg` nor `GRO`, at least as implemented in `quinn-udp`.

The only exception now, with this pull request, is MacOS and iOS with `recvmsg_x`, acting like `recvmmsg`, but not supporting GRO.

File: quinn-udp/src/unix.rs
Comment (mxinden) [reply quinn-udp/src/unix.rs]: Yes. Thank you @larseggert.

**Comment by larseggert** [comment]
Are you waiting on anything from me on this?

File: quinn-udp/src/unix.rs
Comment (mxinden) [inline quinn-udp/src/unix.rs:813]:
```suggestion
// On Apple platforms using the `sendmsg_x` call, UDP datagram segmentation is not
// offloaded to the NIC or even the kernel, but instead done here in user space in
// [`send`]) and then passed to the OS as individual `iovec`s (up to `BATCH_SIZE`).
#[cfg(any(target_os = "macos", target_os = "ios"))]
```

I believe in this codebase convention is to put comments before attributes.

File: quinn-udp/src/unix.rs
Comment (mxinden) [reply quinn-udp/src/unix.rs:46]: > I can put them behind a feature if you prefer.

No strong opinion.

Anyone aware of any other project using these APIs across apple platforms (e.g. MacOS and iOS) and platform versions? If not, I think putting these behind a feature flag, thus allowing for some experimentation first, is the safest way to go.

We can e.g. enable the feature flag on Firefox Nightly, thus at least get to test it on a large set of MacOS machines, before always enabling it.

**Review by mxinden — COMMENTED** [review]
`quinn-udp/benches/throughput.rs` will need more changes to still support non-apple platform. @larseggert I believe we will need to either run it multi-threaded, or use some kind of executor, e.g. `tokio`. I can prepare a commit in the next couple of days. Sorry for missing this in earlier reviews.

Changes itself look good to me.

File: quinn-udp/src/unix.rs
Comment (mxinden) [inline quinn-udp/src/unix.rs:764]:
```suggestion
        target_os = "netbsd",
        target_os = "linux",
```

Also, what about solaris?

File: quinn-udp/src/unix.rs
Comment (mxinden) [inline quinn-udp/src/unix.rs:757]: I think it would be good for `fast-apple-datapath` to be a no-op on non-apple platforms. As it is now, on linux, with `fast-apple-datapath`, this won't compile.

How about something along the lines of:

```
#[cfg(any(
  target_os = // all non apple platforms,
  all(
        not(feature = "fast-apple-datapath"),
        any( /* all apple platoforms */),
  )
))]

File: quinn-udp/benches/throughput.rs
Comment (mxinden) [reply quinn-udp/benches/throughput.rs]: @larseggert will be fixed with https://github.com/larseggert/quinn/pull/1.

Comment (larseggert) [reply quinn-udp/benches/throughput.rs]: Thanks! 

Also, since all the tests are green, I guess quinn is missing a unit test to detect this breakage...

**Comment by djc** [comment]
@Ralith can you do another round on this one?

**Comment by larseggert** [comment]
Once @mxinden's fix to the bench is in, I will squash the PR.

File: quinn-udp/benches/throughput.rs
Comment (mxinden) [reply quinn-udp/benches/throughput.rs]: Above referenced pull request addressing this issue moved to https://github.com/quinn-rs/quinn/pull/2010.

**Comment by AndrewDryga** [comment]
Hey guys 👋, is there any chance Apple won't approve apps that are using those private syscalls in the App Store? They are notorious for doing so and even de-listing apps for using anything "undocumented". See 2.5.1 here: https://developer.apple.com/app-store/review/guidelines/

**Comment by larseggert** [comment]
The use of the private syscalls is now behind a non-default feature.

**Comment by AndrewDryga** [comment]
@larseggert should we add a big fat warning saying that if you enable this flag you will violate Apple ToS so it's only should be enabled if app is not distributed (or notarized for EU)?

**Comment by larseggert** [comment]
I have no opinion here - we don't distribute via the App Store. 

Could you make a suggestion on what you think would be good to add?

**Comment by thomaseizinger** [comment]
> The use of the private syscalls is now behind a non-default feature.

Thank you for making it optional! It can be an issue to toggle code-paths like these with cargo features because they get aggregated across the entire dependency tree. Simply adding another dependency that also happens to depend on `quinn-udp` can thus activate this silently. (As a rule of thumb, cargo features should only add new APIs, not modify existing ones so this doesn't happen.)

For some prior art, `curve25519-dalek` had to solve similar challenges: https://github.com/dalek-cryptography/curve25519-dalek/tree/main/curve25519-dalek#backends

It looks like it is already abstracted away pretty well. Could we change this to a "plain" rustc cfg that needs to be set at build-time?
I am happy for it to be opt-out. Most apps using this probably aren't in the app store and toggling it off is a one-liner in the build process. Ultimately, I got no opinion though on opt-in/opt-out.

**Comment by mxinden** [comment]
> Simply adding another dependency that also happens to depend on `quinn-udp` can thus activate this silently.

Ignoring the case where the activation is unintentional, if that other dependency can use `quinn-udp` with `sendmsg_x` and `recvmsg_x`, I don't see a reason why other consumers of `quinn-udp` within the same binary, shouldn't use `sendmsg_x` and `recvmsg_x`. In other words, I don't think this needs to be optional per consumer of `quinn-udp`, but instead only optional for the entire binary. Thus I think a feature flag suffices.

> Could we change this to a "plain" rustc cfg that needs to be set at build-time?

I might be missing something. Doesn't the same concern apply both to plain cargo features and rustc cfgs? Aren't rustc cfgs also global, as in apply to the entire binary?

If so, I would opt for the, in my eyes, more idiomatic cargo feature instead of the less intuitive rustc cfg.

**Comment by djc** [comment]
> > Could we change this to a "plain" rustc cfg that needs to be set at build-time?
> 
> I might be missing something. Doesn't the same concern apply both to plain cargo features and rustc cfgs? Aren't rustc cfgs also global, as in apply to the entire binary?
> 
> If so, I would opt for the, in my eyes, more idiomatic cargo feature instead of the less intuitive rustc cfg.

The `cfg` flag is set at build time. Cargo features are propagated through the dependency graph so that if you add a dependency to your binary that has a feature-enabled dependency on one of your dependencies, that feature now also gets enabled for your project. This can "silently" happen without you noticing, so it's a bit of a footgun.

**Comment by mxinden** [comment]
> The `cfg` flag is set at build time.

I don't understand why this is relevant. Both cargo features and rustc cfgs happen at build time, i.e. not at runtime.

> Cargo features are propagated through the dependency graph so that if you add a dependency to your binary that has a feature-enabled dependency on one of your dependencies, that feature now also gets enabled for your project.

Correct. My argument above is, that this isn't a problem here. I argue that if it is safe for one consumer of `quinn-udp` to use `sendmsg_x` and `recvmsg_x`, it is safe for all consumers of `quinn-udp` within the same binary to use `sendmsg_x` and `recvmsg_x`. 

Can you think of a scenario where that is not the case?

>  This can "silently" happen without you noticing, so it's a bit of a footgun.

In this particular case, if you don't trust your dependency to make wise choices on their `quinn-udp` feature selection, I don't think one should use that dependency in the first place.

> I am happy for it to be opt-out.

I suggest for it to be opt-in for now, given that we have little to no experience with these APIs, nor are we aware of any other applications using these APIs.

**Comment by djc** [comment]
> > The `cfg` flag is set at build time.
> 
> I don't understand why this is relevant. Both cargo features and rustc cfgs happen at build time, i.e. not at runtime.

Because `cfg` flags are set from the command/environment where the end artifact is built, whereas Cargo features can be set by other dependencies which often aren't under your control.

> > Cargo features are propagated through the dependency graph so that if you add a dependency to your binary that has a feature-enabled dependency on one of your dependencies, that feature now also gets enabled for your project.
> 
> Correct. My argument above is, that this isn't a problem here. I argue that if it is safe for one consumer of `quinn-udp` to use `sendmsg_x` and `recvmsg_x`, it is safe for all consumers of `quinn-udp` within the same binary to use `sendmsg_x` and `recvmsg_x`.
> 
> Can you think of a scenario where that is not the case?
> 
> > This can "silently" happen without you noticing, so it's a bit of a footgun.
> 
> In this particular case, if one doesn't trust one's dependency to make wise choices on their `quinn-udp` feature selection, I don't think one should use that dependency in the first place.

I don't think we're going to solve the crates.io supply chain issues here, and I think not everyone is as careful as Mozilla as vetting their dependencies (and any updates happening to them over time). As such, IMO it is a best practice that additional functionality enabled via features should additionally be toggled by API choices. On the other hand, (a) I'm not sure how ergonomic that would be in this case, (b) I don't like the ergonomics of `cfg` as an alternative.

> > I am happy for it to be opt-out.
> 
> I suggest for it to be opt-in for now, given that we have little to no experience with these APIs, nor are we aware of any other applications using these APIs.

This seems reasonable to me.

**Comment by mxinden** [comment]
> > > The `cfg` flag is set at build time.
> > 
> > 
> > I don't understand why this is relevant. Both cargo features and rustc cfgs happen at build time, i.e. not at runtime.
> 
> Because `cfg` flags are set from the command/environment where the end artifact is built, whereas Cargo features can be set by other dependencies which often aren't under your control.

Thank you for expanding on this. That makes sense.

**Comment by larseggert** [comment]
See 4f86c5058740a9295b74f87c8cab6206a60e9971. Let me know if I misunderstood something.

File: quinn-udp/src/unix.rs
Comment (Ralith) [inline quinn-udp/src/unix.rs:489]:
```suggestion
                max_msg_count as _,
```

**Review by Ralith — APPROVED** [review]

**Comment by djc** [comment]
> See [4f86c50](https://github.com/quinn-rs/quinn/commit/4f86c5058740a9295b74f87c8cab6206a60e9971). Let me know if I misunderstood something.

Discussed this with @Ralith privately, we feel that the use of `cfg` has a negative impact on the ergonomics and the chance of issues arising from silent feature enabling is pretty limited in this case, so would prefer to stick with a feature for this for now.

**Comment by djc** [comment]
Sorry it took a while to get the benchmark PR merged, can you rebase on top of main?

**Comment by larseggert** [comment]
Shouldn't the `features` CI workflow have included `fast-apple-datapath`? I don't see it in the logs. (Or I am missing something.)

**Comment by mxinden** [comment]
Should be fine I believe:

```
running `cargo check --no-default-features --features direct-log,fast-apple-datapath,log,tracing` on quinn-udp (1754/1772)
running `cargo check --no-default-features --features default` on quinn-udp (1755/1772)
running `cargo check --no-default-features --features direct-log` on quinn-udp (1756/1772)
running `cargo check --no-default-features --features default,direct-log` on quinn-udp (1757/1772)
running `cargo check --no-default-features --features fast-apple-datapath` on quinn-udp (1758/1772)
running `cargo check --no-default-features --features default,fast-apple-datapath` on quinn-udp (1759/1772)
running `cargo check --no-default-features --features direct-log,fast-apple-datapath` on quinn-udp (1760/1772)
running `cargo check --no-default-features --features default,direct-log,fast-apple-datapath` on quinn-udp (1761/1772)
running `cargo check --no-default-features --features log` on quinn-udp (1762/1772)
running `cargo check --no-default-features --features direct-log,log` on quinn-udp (1763/1772)
running `cargo check --no-default-features --features fast-apple-datapath,log` on quinn-udp (1764/1772)
running `cargo check --no-default-features --features direct-log,fast-apple-datapath,log` on quinn-udp (1765/1772)
running `cargo check --no-default-features --features tracing` on quinn-udp (1766/1772)
running `cargo check --no-default-features --features direct-log,tracing` on quinn-udp (1767/1772)
running `cargo check --no-default-features --features fast-apple-datapath,tracing` on quinn-udp (1768/1772)
running `cargo check --no-default-features --features direct-log,fast-apple-datapath,tracing` on quinn-udp (1769/1772)
running `cargo check --no-default-features --features log,tracing` on quinn-udp (1770/1772)
running `cargo check --no-default-features --features direct-log,log,tracing` on quinn-udp (1771/1772)
running `cargo check --no-default-features --features fast-apple-datapath,log,tracing` on quinn-udp (1772/1772)
```

File: quinn-udp/src/unix.rs
Comment (mxinden) [reply quinn-udp/src/unix.rs:376]: Agreed. @larseggert I will follow-up with a pull request.

**Comment by mxinden** [comment]
For those interested in using the `fast-apple-datapath` feature:
- All MacOS Firefox Nightly users [are now using it](https://phabricator.services.mozilla.com/D234901).
- We see [~11% performance improvement](https://github.com/mozilla/neqo/pull/2302) on our localhost CPU bound throughput benchmark.
- Every other (50th percentile) QUIC UDP `recvmsg_x` call reads 2 datagrams.


image

https://glam.telemetry.mozilla.org/fog/probe/networking_http_3_udp_datagram_segments_received/explore?aggType=avg&os=Darwin&timeHorizon=WEEK&visiblePercentiles=%5B99.9%2C99%2C95%2C75%2C50%2C25%2C5%5D

**Pull request merged**
